### PR TITLE
Add support for inputs

### PR DESCRIPTION
**Pull request by Manishearth**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,6 +11,7 @@ script:
   - cargo fmt --all -- --check
   - cd webxr; cargo build --features=glwindow
   - cd webxr; cargo build --features=ipc,glwindow
+  - cd webxr; cargo build --features=ipc,headless
```
Comment (asajeffrey) [inline .travis.yml]: Can just make this `--features=glwindow,headless` and `--features=ipc,glwindow,headless`

File: webxr-api/device.rs [hunk 2]
```diff
@@ -40,4 +41,6 @@ pub trait Device: 'static {
     /// While this method is being called, the device has unique access
     /// to the texture. The texture should be sync'd using glWaitSync before being used.
     fn render_animation_frame(&mut self, texture_id: u32, size: Size2D<i32>, sync: GLsync);
+
+    fn initial_inputs(&self) -> Vec<InputInfo>;
```
Comment (asajeffrey) [inline webxr-api/device.rs]: Bikeshedding: not sure about the name. `input_sources: Vec<InputSource>`? Needs some documentation too.

File: webxr-api/input.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    TrackedPointer,
+    Screen,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "ipc", derive(serde::Serialize, serde::Deserialize))]
+pub struct InputInfo {
+    pub handedness: Handedness,
+    pub target_ray_mode: TargetRayMode,
+    pub id: u32,
```
Comment (asajeffrey) [inline webxr-api/input.rs]: Oh joy, ids. Can we at least declare a type synonym `InputId`?

File: webxr-api/input.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::Input;
+use crate::Native;
+
+use euclid::TypedRigidTransform3D;
+
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "ipc", derive(serde::Serialize, serde::Deserialize))]
+pub enum Handedness {
```
Comment (asajeffrey) [inline webxr-api/input.rs]: Spec link to https://immersive-web.github.io/webxr/#enumdef-xrhandedness

File: webxr-api/input.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    None,
+    Left,
+    Right,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "ipc", derive(serde::Serialize, serde::Deserialize))]
+pub enum TargetRayMode {
```
Comment (asajeffrey) [inline webxr-api/input.rs]: Spec link to https://immersive-web.github.io/webxr/#enumdef-xrtargetraymode

File: webxr-api/frame.rs [hunk 6]
```diff
@@ -18,4 +19,6 @@ pub struct Frame {
     /// This is equivalent to the pose of the viewer in native coordinates.
     /// This is the inverse of the view matrix.
     pub transform: TypedRigidTransform3D<f32, Viewer, Native>,
+
+    pub inputs: Vec<InputFrame>,
```
Comment (asajeffrey) [inline webxr-api/frame.rs]: Bikeshed: Should this be `events: Vec<Event>`? or `Vec<(InputSourceId, InputEvent)>` or some such?

File: webxr-api/input.rs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    Gaze,
+    TrackedPointer,
+    Screen,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "ipc", derive(serde::Serialize, serde::Deserialize))]
+pub struct InputInfo {
```
Comment (asajeffrey) [inline webxr-api/input.rs]: Bikeshed: should this be `InputSource`?

File: webxr-api/input.rs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    pub handedness: Handedness,
+    pub target_ray_mode: TargetRayMode,
+    pub id: u32,
+}
+
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "ipc", derive(serde::Serialize, serde::Deserialize))]
+pub struct InputFrame {
```
Comment (asajeffrey) [inline webxr-api/input.rs]: We're going to want more of these as time goes no, so make this an `enum InputEvent`?

File: webxr-api/device.rs (hunk 2 above)
Comment (Manishearth) [reply webxr-api/device.rs]: This is about the initial input sources only, we don't handle changes yet. The client will have to keep track of changes.

File: webxr-api/input.rs (hunk 3 above)
Comment (Manishearth) [reply webxr-api/input.rs]: The spec actually exposes these :smile: . It's part of gamepad but also used in input.

File: webxr-api/frame.rs (hunk 6 above)
Comment (Manishearth) [reply webxr-api/frame.rs]: I don't want to think about events right now, trying to implement the minimum possible for integration. Also events can be triggered without an rAF loop, so we can't stuff that in the frame.

File: webxr-api/input.rs (hunk 8 above)
Comment (Manishearth) [reply webxr-api/input.rs]: No, events should not be handled here

File: webxr-api/device.rs (hunk 2 above)
Comment (asajeffrey) [reply webxr-api/device.rs]: Still needs documentation!

File: webxr-api/input.rs (hunk 3 above)
Comment (asajeffrey) [reply webxr-api/input.rs]: Oh yuk, not even with a type synonym?

File: webxr-api/frame.rs (hunk 6 above)
Comment (asajeffrey) [reply webxr-api/frame.rs]: How about `input_poses: Vec<TypedTransform3D<f32, Input, Native>>` then?

File: webxr-api/input.rs (hunk 8 above)
Comment (asajeffrey) [reply webxr-api/input.rs]: OKI, but if this is just for pose info then just using `TypedTransform3D<f32, Input, Native>` seems simplest.

File: webxr-api/frame.rs (hunk 6 above)
Comment (Manishearth) [reply webxr-api/frame.rs]: It's going to eventually also contain gaze info, and gamepad info

File: webxr-api/input.rs (hunk 3 above)
Comment (Manishearth) [reply webxr-api/input.rs]: I added one.

File: webxr-api/frame.rs (hunk 6 above)
Comment (asajeffrey) [reply webxr-api/frame.rs]: At this point I am happy with whatever you want to do. A comment would be nice though!

**Comment by asajeffrey** [comment]
It could do with a few more code comments, but otherwise you can r=me.

**Pull request merged**